Inductor cache: Revamp how we handle frozen params

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143808

Summary: In https://github.com/pytorch/pytorch/pull/143563 we have a report of a problem with the treatment of frozen params in the inductor cache implementation. There seems to be a path where new constants are added in the `GraphLowering`. On a cache hit when we try to find those constant names in the `torch.fx.GraphModule`, they do not exist. The current approach treats all constants differently if the GM has any frozen params. This PR changes the approach to only treat the _frozen_ params specially, but store all other constants in the cache entry (as we do without freezing):
1) When creating a cache entry, store the names of any frozen params, but the values of any other constants.
2) On a cache hit, restore the values of the frozen params by looking up in the current GM. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

masnesral committed:
[RFC] Inductor cache: Revamp how we handle frozen params

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] Inductor cache: Revamp how we handle frozen params"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -551,40 +557,43 @@ def _reduce_fake_tensor(
-        self,
-        t: Tensor,
-    ) -> Tuple[Callable[[T], T], Tuple[Union[TensorMetadata, TensorMetadataAndValues]]]:
+        self, t: Tensor
```
The decision to consider the metadata or metadata+values happens in `_reduce_parameter` now, so this method always includes the values

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -551,40 +557,43 @@ def _reduce_fake_tensor(
+        return (_ident, (TensorMetadataAndValues(metadata, values),))
 
-            return (_ident, (TensorMetadataAndValues(metadata, values),))
+    def _reduce_parameter(
```
The assumption here is that for hashing purposes, I can consider Parameters to be Tensors

leslie-fang-intel commented:
Thanks for the fix to: https://github.com/pytorch/pytorch/issues/143144

eellison reviewed (commented):
The parameters thing should mostly work, but wouldn't handle say Paremeters which have become subclasses. We can land this, but would be more general to record newly added constants that occur in lowering ?

I'm open to either solution .

leslie-fang-intel commented:
Hi @masnesral, is this PR ready to land?

masnesral committed:
Update on "[RFC] Inductor cache: Revamp how we handle frozen params"


Summary: In https://github.com/pytorch/pytorch/pull/143563 we have a report of a problem with the treatment of frozen params in the inductor cache implementation. There seems to be a path where new constants are added in the `GraphLowering`. On a cache hit when we try to find those constant names in the `torch.fx.GraphModule`, they do not exist. The current approach treats all constants differently if the GM has any frozen params. This PR changes the approach to only treat the _frozen_ params specially, but store all other constants in the cache entry (as we do without freezing):
1) When creating a cache entry, store the names of any frozen params, but the values of any other constants.
2) On a cache hit, restore the values of the frozen params by looking up in the current GM. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] Inductor cache: Revamp how we handle frozen params"


Summary: In https://github.com/pytorch/pytorch/pull/143563 we have a report of a problem with the treatment of frozen params in the inductor cache implementation. There seems to be a path where new constants are added in the `GraphLowering`. On a cache hit when we try to find those constant names in the `torch.fx.GraphModule`, they do not exist. The current approach treats all constants differently if the GM has any frozen params. This PR changes the approach to only treat the _frozen_ params specially, but store all other constants in the cache entry (as we do without freezing):
1) When creating a cache entry, store the names of any frozen params, but the values of any other constants.
2) On a cache hit, restore the values of the frozen params by looking up in the current GM. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[RFC] Inductor cache: Revamp how we handle frozen params"


Summary: In https://github.com/pytorch/pytorch/pull/143563 we have a report of a problem with the treatment of frozen params in the inductor cache implementation. There seems to be a path where new constants are added in the `GraphLowering`. On a cache hit when we try to find those constant names in the `torch.fx.GraphModule`, they do not exist. The current approach treats all constants differently if the GM has any frozen params. This PR changes the approach to only treat the _frozen_ params specially, but store all other constants in the cache entry (as we do without freezing):
1) When creating a cache entry, store the names of any frozen params, but the values of any other constants.
2) On a cache hit, restore the values of the frozen params by looking up in the current GM. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

sanchitintel commented:
Hi, can this PR be merged with `-f SOME_EXPLANATION`? Thanks!

leslie-fang-intel commented:
> @leslie-fang-intel, the bug that this PR fixes is adversely affecting some workloads that use multiple PyTorch processes with max-autotune on the same machine. This bug is also present in `release/2.6` branch, so could we cherry-pick this commit to `release/2.6`, when it lands?

I also prefer a cherry-pick back to `release/2.6` branch since it fixes crash issue as reported in https://github.com/pytorch/pytorch/issues/143144. But since it's pretty close to the final release, I am not sure if it's still applicable.

masnesral commented:
@pytorchbot merge

eellison commented:
cc @masnesral this is leading to a big compile time regression with freezing. see `python benchmarks/dynamo/torchbench.py --backend inductor --device cuda --only alexnet --bfloat16 --cold-start-latency --print-compilation-time --inference --performance --freezing` goes from 5 seconds to 20 seconds. 

And this more comprehensive dashboard results: https://hud.pytorch.org/benchmark/torchbench/inductor_with_cudagraphs_freezing?dashboard=torchinductor&startTime=Mon,%2020%20Jan%202025%2019:15:46%20GMT&stopTime=Mon,%2027%20Jan%202025%2019:15:46%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cuda%20(a100)&lBranch=main&lCommit=c5216d2b6c6613ff931c02eca4f840549fe37e62&rBranch=main&rCommit=3cbc8c54fd37eb590e2a9206aecf3ab568b3e63c

masnesral commented:
cc @leslie-fang-intel re: your comments about cherry-pick above. See Elias's comment. This looks like a pretty bad regression :(  I'm working on a fix

masnesral commented:
@leslie-fang-intel , here's the fix: https://github.com/pytorch/pytorch/pull/145868
